File: app/src/main/java/br/com/menuvem/lojista/presentation/products/ProdutoFormDialog.kt
package br.com.menuvem.lojista.presentation.products

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import br.com.menuvem.lojista.domain.model.Produto

/**
 * Dialog de criação/edição de produto: nome, margem-alvo (%) e preço de
 * venda atual (opcional — usado para comparar a margem real com a meta).
 */
@Composable
fun ProdutoFormDialog(
    produto: Produto? = null,
    onDismiss: () -> Unit,
    onConfirm: (nome: String, margemAlvo: Double, precoVenda: Double?) -> Unit
) {
    var nome by remember { mutableStateOf(produto?.nome ?: "") }
    var margem by remember {
        mutableStateOf(
            produto?.margemAlvoPercentual?.let { m ->
                if (m % 1.0 == 0.0) m.toInt().toString() else m.toString().replace('.', ',')
            } ?: "30"
        )
    }
    var preco by remember {
        mutableStateOf(produto?.precoVendaAtual?.let { String.format("%.2f", it).replace('.', ',') } ?: "")
    }

    val margemVal = margem.replace(",", ".").toDoubleOrNull()
    val precoVal = preco.replace(",", ".").toDoubleOrNull()
    val isValid = nome.isNotBlank() &&
            margemVal != null && margemVal in 0.0..99.9 &&
            (preco.isBlank() || precoVal != null)

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(if (produto == null) "Novo Produto" else "Editar Produto") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                OutlinedTextField(
                    value = nome,
                    onValueChange = { nome = it },
                    label = { Text("Nome") },
                    placeholder = { Text("Ex.: Pizza Grande de Calabresa") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Sentences),
                    modifier = Modifier.fillMaxWidth()
                )
                OutlinedTextField(
                    value = margem,
                    onValueChange = { margem = it },
                    label = { Text("Margem de lucro alvo (%)") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
                    modifier = Modifier.fillMaxWidth()
                )
                OutlinedTextField(
                    value = preco,
                    onValueChange = { preco = it },
                    label = { Text("Preço de venda atual (opcional)") },
                    placeholder = { Text("0,00") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
                    modifier = Modifier.fillMaxWidth()
                )
            }
        },
        confirmButton = {
            TextButton(
                onClick = {
                    onConfirm(nome.trim(), margemVal!!, if (preco.isBlank()) null else precoVal)
                },
                enabled = isValid
            ) { Text("Salvar") }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text("Cancelar") }
        }
    )
}

File: lib/presentation/products/produto_form_dialog.dart
import 'package:flutter/material.dart';

import '../../domain/model/produto.dart';
import '../../theme/app_theme.dart';
import '../components/emoji_picker_field.dart';
import '../components/emojis.dart';
import '../components/formatters.dart';

typedef ProdutoFormConfirm = void Function(
  String nome,
  double margemAlvo,
  double? precoVenda,
  String? emoji,
);

/// Folha de criação/edição de produto: nome, margem-alvo (%) e preço de
/// venda atual (opcional — usado para comparar a margem real com a meta).
///
/// O botão Salvar fica no cabeçalho fixo (nunca atrás do teclado) — só o
/// conteúdo abaixo rola.
class ProdutoFormDialog extends StatefulWidget {
  const ProdutoFormDialog({super.key, this.produto, required this.onConfirm});

  final Produto? produto;
  final ProdutoFormConfirm onConfirm;

  @override
  State<ProdutoFormDialog> createState() => _ProdutoFormDialogState();
}

class _ProdutoFormDialogState extends State<ProdutoFormDialog> {
  late final _nomeController = TextEditingController(text: widget.produto?.nome ?? '');
  late final _margemController = TextEditingController(
    text: widget.produto == null
        ? '30'
        : (widget.produto!.margemAlvoPercentual % 1.0 == 0.0
            ? widget.produto!.margemAlvoPercentual.toInt().toString()
            : widget.produto!.margemAlvoPercentual.toString().replaceAll('.', ',')),
  );
  late final _precoController = TextEditingController(
    text: widget.produto?.precoVendaAtual != null
        ? formatarMoeda(widget.produto!.precoVendaAtual!)
        : '',
  );
  late String? _emoji = widget.produto?.emoji;

  @override
  void dispose() {
    _nomeController.dispose();
    _margemController.dispose();
    _precoController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final margemVal = parseDecimalPtBr(_margemController.text);
    final precoVal = parseDecimalPtBr(_precoController.text);
    final isValid = _nomeController.text.trim().isNotEmpty &&
        margemVal != null &&
        margemVal >= 0 &&
        margemVal <= 99.9 &&
        (_precoController.text.trim().isEmpty || precoVal != null);

    return Padding(
      padding: EdgeInsets.only(bottom: MediaQuery.of(context).viewInsets.bottom),
      child: SingleChildScrollView(
        padding: const EdgeInsets.fromLTRB(20, 12, 20, 20),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Center(
              child: Container(
                width: 40,
                height: 4,
                decoration: BoxDecoration(
                  color: outlineLight,
                  borderRadius: BorderRadius.circular(4),
                ),
              ),
            ),
            const SizedBox(height: 12),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                TextButton(
                  onPressed: () => Navigator.of(context).pop(),
                  child: const Text('Cancelar'),
                ),
                Text(widget.produto == null ? 'Novo Produto' : 'Editar Produto',
                    style: Theme.of(context)
                        .textTheme
                        .titleMedium
                        ?.copyWith(fontWeight: FontWeight.w600)),
                TextButton(
                  onPressed: isValid
                      ? () {
                          widget.onConfirm(
                            _nomeController.text.trim(),
                            margemVal,
                            _precoController.text.trim().isEmpty ? null : precoVal,
                            _emoji,
                          );
                          Navigator.of(context).pop();
                        }
                      : null,
                  child: const Text('Salvar'),
                ),
              ],
            ),
            const SizedBox(height: 8),
            TextField(
              controller: _nomeController,
              decoration: const InputDecoration(
                labelText: 'Nome',
                hintText: 'Ex.: Pizza Grande de Calabresa',
              ),
              textCapitalization: TextCapitalization.sentences,
              onChanged: (_) => setState(() {}),
            ),
            const SizedBox(height: 12),
            Text('Ícone (opcional)', style: Theme.of(context).textTheme.labelMedium),
            const SizedBox(height: 8),
            EmojiPickerField(
              opcoes: emojisProduto,
              selecionado: _emoji,
              onChanged: (v) => setState(() => _emoji = v),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _margemController,
              decoration: const InputDecoration(labelText: 'Margem de lucro alvo (%)'),
              keyboardType: const TextInputType.numberWithOptions(decimal: true),
              onChanged: (_) => setState(() {}),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _precoController,
              decoration: const InputDecoration(
                labelText: 'Preço de venda atual (opcional)',
                hintText: '0,00',
              ),
              keyboardType: const TextInputType.numberWithOptions(decimal: true),
              onChanged: (_) => setState(() {}),
            ),
          ],
        ),
      ),
    );
  }
}
